File: C++/LinkedList/LinkedList.cpp
#include "LinkedList.h"

template<typename T>
LinkedList<T>::LinkedList() : 
    m_head(nullptr), m_tail(nullptr), m_length(0) {}

template<typename T>
LinkedList<T>::LinkedList(const LinkedList<T>& other) {
    
    if (other.size() == 0) {
        m_head = nullptr;
        m_tail = nullptr;
        m_length = 0;
    } else {
        ListNode<T>* currentOther = other.head();
        m_head = (new ListNode<T>(currentOther->get()));
        ListNode<T>* currentThis = m_head;
        while(currentOther->next()) {
            currentOther = currentOther->next();
            currentThis->setNext(
                new ListNode<T>(currentOther->get()));
            currentThis = currentThis->next();
        }
        m_tail = currentThis;
        m_length = other.size();
    }
}

template<typename T>
LinkedList<T>::~LinkedList() {
    if(m_head) {
        ListNode<T>* ptr = m_head;
        while(ptr) {
            ListNode<T>* tmp = ptr;
            ptr = tmp->next();
            delete tmp;
        }
    }
}

template<typename T>
LinkedList<T>& LinkedList<T>::operator=(const LinkedList<T>& other) {
    if (this == &other) return *this;
    if (m_head) {
        ListNode<T>* ptr = m_head;
        while(ptr) {
            ListNode<T>* tmp = ptr;
            ptr = tmp->next();
            delete tmp;
        }
        m_head = nullptr;
        m_tail = nullptr;
        m_length = 0;
    }
    if (other.head()) {
        ListNode<T>* currentOther = other.head();
        m_head = new ListNode<T>(currentOther->get());
        ListNode<T>* currentThis = m_head;
        while (currentOther->next()) {
            currentOther = currentOther->next();
            currentThis->setNext(
                new ListNode<T>(currentOther->get()));
            currentThis = currentThis->next();
        }
        m_tail = currentThis;
        m_length = other.size();
    } else {
        m_head = nullptr;
        m_tail = nullptr;
        m_length = 0;
    }
    return *this;
}

template<typename T>
T LinkedList<T>::get(int index) {
    if (m_length == 0 || index >= m_length || index < 0) 
        throw std::out_of_range("Index out of range of list object");
    int i = 0;
    ListNode<T>* tmp;
    tmp = m_head;
    while(i < index) {
        tmp = tmp->next();
        ++i;
    }
    return tmp->get();
}

template<typename T>
size_t LinkedList<T>::size() const { return m_length; }

template<typename T>
ListNode<T>* LinkedList<T>::head() const { return m_head; }

template<typename T>
ListNode<T>* LinkedList<T>::tail() const { return m_tail; }

template<typename T>
void LinkedList<T>::pushBack(T val) {
    if (m_head && m_tail) {
        m_tail->setNext(new ListNode<T>(val));
        m_tail = m_tail->next();
    } else {
        m_head = new ListNode<T>(val);
        m_tail = m_head;
    }
    ++m_length;
}

template<typename T>
void LinkedList<T>::pushFront(T val) {
    if (m_head && m_tail) {
        ListNode<T>* tmp = new ListNode<T>(val);
        tmp->setNext(m_head);
        m_head = tmp;
    } else {
        m_head = new ListNode<T>(val);
        m_tail = m_head;
    }
    ++m_length;
}

template class LinkedList<int>;

File: python/LinkedList.py
from ListNode import ListNode

class LinkedList:
    head: ListNode
    tail: ListNode
    length: int
    
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        
    def push(self, val):
        if self.head == None:
            self.head = ListNode(val)
            self.tail = self.head
        else:
            self.tail.setNext(ListNode(val))
            self.tail = self.tail.next
        self.length += 1
        
    def pop(self):
        if self.tail == None: raise LookupError
        current = self.tail
        if self.tail == self.head:
            self.head = None
            self.tail = None
            self.length = 0
        else:
            newTail = self.head
            while newTail.next != current:
                newTail = newTail.next
            newTail.setNext(None)
            self.tail = newTail
            self.length -= 1
        return current.value
            
    
    # get the item using square brackets
    def __getitem__(self, index: int):
        if index < 0 or index >= self.length: raise IndexError
        current = self.head
        for x in range(0, index):
            current = current.next
        return current.value
    
    def __setitem__(self, index: int, val):
        if index < 0 or index >= self.length: raise IndexError
        current = self.head
        for x in range(0, index):
            current = current.next
        current.value = val